third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp

```diff
@@ -772,6 +846,390 @@ LogicalResult Pingponger::sliceDot(OpBuilder &builder, Location loc,
   return success();
 }
 
+LogicalResult Pingponger::genAsyncCopySlices(OpBuilder &builder) {
+  if (asyncCopyOps.empty())
+    return success();
+
+  auto &loopBody = forOp.getRegion().front();
+  auto yieldOp = cast<scf::YieldOp>(loopBody.getTerminator());
+
+  for (auto asyncCopy : asyncCopyOps) {
+    MLIRContext *ctx = asyncCopy->getContext();
+    auto srcPointers = asyncCopy.getSrc();
+    auto subView = cast<triton::gpu::MemDescSubviewOp>(
+        asyncCopy.getResult().getDefiningOp());
+    auto subViewEncoding = subView.getType().getEncoding();
+
+    DenseMap<int64_t, triton::gpu::MemDescSubviewOp> subViews;
+    for (auto user : subView->getUsers()) {
+      if (auto subView = dyn_cast<triton::gpu::MemDescSubviewOp>(user)) {
+        if (auto attr = subView->getAttrOfType<mlir::IntegerAttr>(
+                Pingponger::sliceAttrName)) {
+          auto sliceIdx = attr.getValue().getSExtValue();
+          subViews.insert({sliceIdx, subView});
+
+          if (!newAsyncGroups.contains(sliceIdx)) {
+            newAsyncGroups.insert({sliceIdx, {}});
+          }
+        }
+      }
+    }
+
+    if (subViews.empty()) {
+      auto commit =
+          getSingleUserOf<ttg::AsyncCommitGroupOp>(asyncCopy.getToken());
+      preservedAsyncCommits.insert(commit->getResult());
+      continue;
+    }
+
+    // infer the sliced shape
+    triton::gpu::MemDescSubviewOp subViewSlice = subViews[0];
+    auto origShape = subView.getType().getShape();
+    auto slicedShape = subViewSlice.getType().getShape();
+    assert(origShape.size() == slicedShape.size());
+    const auto numDims = origShape.size();
+    int64_t slicedDim = -1;
+    for (size_t dim = 0; dim < numDims; ++dim) {
+      if (origShape[dim] != slicedShape[dim]) {
+        slicedDim = dim;
+        break;
+      }
+    }
+
+    builder.setInsertionPointAfter(asyncCopy);
+
+    auto elementType = srcPointers.getType().getElementType();
+    auto encoding =
+        cast<ttg::BlockedEncodingAttr>(srcPointers.getType().getEncoding());
+
+    auto warpsPerCTA = encoding.getWarpsPerCTA();
+    auto sizePerThread = encoding.getSizePerThread();
+    SmallVector<unsigned> threadPerWarp(warpsPerCTA.size(), 0);
+    for (size_t dim = 0; dim < numDims; ++dim) {
+      threadPerWarp[dim] =
```

Comment:
I have a silly Q for educational purposes, looks like we are only updating the threadsPerWarp here, does it not make more sense to updated the sizePerThread in the newEncoding?

Reply:
No, we want to preserve the number of elements per thread (i.e., each holds `16` consecutive elements of a tensor). We just want to change the layout of threads; thus, to change which part of a tensor is held by CTA.

Each CTA holds `64x128` tile of a tensor in the following example 

```mlir
#blocked = #ttg.blocked<{sizePerThread = [1, 16], threadsPerWarp = [8, 8], warpsPerCTA = [8, 1], order = [1, 0]}>
```
But if we change `threadsPerWarp` to `[32,2]` then a CTA holds `256x32` tile of a tensor. 

```mlir
#blocked3 = #ttg.blocked<{sizePerThread = [1, 16], threadsPerWarp = [32, 2], warpsPerCTA = [8, 1], order = [1, 0]}>
```

In both cases, a thread holds 16 consecutive elements which determines the width of load instructions

Reply:
Thanks for the answer, that's really interesting! A follow up question, if we change the which part of tensor the CTA will hold, wouldn't we need to have an extra global read (or potentially fused but wider global read) to get that part of tensor?

Reply:
ConverLayout results in some machine ops if it cannot be fully optimized (propagated till the top of a function). If I remember correctly, the layout change is going to happen in LDS

Reply:
Nice! didn't realize this was pre-layout propagation phase, but this makes a lot of sense, thanks :)